README.md
# Docker Mattermost

This image provides the [Mattermost (Team Edition)](https://www.mattermost.org)
chat application server as a Docker image.


# Upgrade notes

Since version 3.8 Mattermost supports setting configuration values via
environment variables. As this perfectly matches with common container
practices and allows to remove the logic to write a configuration file from
environment variables during startup, the former environment variable and their
resolution have been removed. This increases maintainability of this image and
keeps it tighter to Mattermost itself.

*This means that you need to redefine any environment variable that you used
before the 3.8 release of this image. See `CHANGELOG.md` for a reference
regarding some important variables.*


# Quick Start

The quickest way to get started is using [docker-compose](https://docs.docker.com/compose/).

Copy the [`docker-compose.yml`](https://raw.githubusercontent.com/telota/docker-mattermost/master/docker-compose.yml)
file, adjust as needed and start the project's containers:

    docker-compose up


You can generate the random strings for salts &ndash; though this isn't needed for testing purposes &ndash; with:

    pwgen -s1 32


With the provided example configuration, open http://localhost:8080 in a
browser and create your administrator account.

You should now have the Mattermost application up and ready for testing. If you want to use this image in production the please read on.


# Configuration

## Configuration file and environment variables

This image comes with the [default configuration file](https://github.com/mattermost/platform/blob/master/config/config.json)
which you may replace by mounting your customized one into
`/opt/mattermost/mattermost/config`.

It is however recommended to override setting with environment variables.
For the defaults see the config file linked above. Please refer to
[this documentation](https://docs.mattermost.com/administration/config-settings.html)
regarding naming of environment variables and available configuration options.

**MAKE SURE THAT ANY OF YOUR DEPLOYMENTS IS CONFIGURED WITH UNIQUE VALUES PER
INSTANCE FOR EACH OF THESE SETTINGS:**

(noted by their environment variable name here)

- `MM_EMAILSETTINGS_INVITESALT`
- `MM_EMAILSETTINGS_PASSWORDRESETSALT`
- `MM_SQLSETTINGS_ATRESTENCRYPTKEY`

Furthermore you need to set `MM_SERVICESETTINGS_SITEURL`.

You may set `DEBUG` to `true` for a verbose bootstrapping and to set
Mattermost's debug level to `DEBUG`.

## File Assets Store

Mattermost stores data in the file system for assets like file uploads and avatars.
The image defines a [volume](https://docs.docker.com/engine/tutorials/dockervolumes/)
for that location.

## Database

Mattermost uses a database backend to store most of its data.
You can configure this image to use MySQL or Postgres.

### Autodiscovery

The simplest configuration uses a container dedicated as database backend, like
in the example `docker-compose.yml`.
As the service names (`postgres` or `mysql`) will be used as hostnames, these
will be probed and any configuration values will be derived from it and use
default values. Just make sure that `DB_PASS` in the `mattermost` service and
`POSTGRES_PASSWORD` resp. `MYSQL_ROOT_PASSWORD` in the database service are
identical.

### Datasource URL

You can also define the database configuration as URL with the
`MM_SQLSETTINGS_DATASOURCE` and an according `MM_SQLSETTINGS_DRIVERNAME`
environment variable.

### Granular settings variables

Lastly the `MM_SQLSETTINGS_DATASOURCE` variable can be assembled from these
variables:

- `MM_SQLSETTINGS_DRIVERNAME` (determines postgres / mysql defaults)
- `DB_ENCODING` (default: none / `utf8mb4,utf8`)
- `DB_HOST`
- `DB_NAME` (default: `mattermost`)
- `DB_PARAMS` (default: `sslmode=disable&connect_timeout=10` / `charset=${DB_ENCODING}`)
- `DB_PASS`
- `DB_PORT` (default: `5432` / `3306`)
- `DB_USER` (default: `postgres` / `root`)


# Building the image

The source repository contains a `Makefile` that can be used to build the image
locally:

    make


# References

- [Source repository of this image](https://github.com/telota/docker-mattermost)
- [Image repository on the Docker Hub](https://hub.docker.com/r/telota/mattermost-team-edition/)

- [Mattermost changelog](https://docs.mattermost.com/administration/changelog.html)
- [Mattermnost source repository](https://github.com/mattermost/platform)
- [Mattermost documentation](https://docs.mattermost.com)


CHANGELOG.md
# Changelog

This file only reflects the changes that are made in this image. Please refer to
the Mattermost [CHANGELOG](http://docs.mattermost.com/administration/changelog.html)
for the list of changes in Mattermost.

## 3.8.0

- mattermost 3.8.0
- switches to Mattermost's environment variables for configuration, replace
  the names in your configuration like this:
  - `MATTERMOST_SECRET_KEY` -> `MM_SQLSETTINGS_ATRESTENCRYPTKEY`
  - `MATTERMOST_LINK_SALT` -> `MM_FILESETTINGS_PUBLICLINKSALT`
  - `MATTERMOST_RESET_SALT` -> `MM_EMAILSETTINGS_PASSWORDRESETSALT`
  - `MATTERMOST_INVITE_SALT` -> `MM_EMAILSETTINGS_INVITESALT`

## 3.7.3

- mattermost 3.7.3

## 3.6.2

- mattermost 3.6.2
- add `MATTERMOST_LOG_CONSOLE_LEVEL`, `MATTERMOST_LOG_FILE_LEVEL` and
  `MATTERMOST_LOG_FILE_FORMAT`

## 3.5.1

- mattermost 3.5.1

## 3.4.0-1

- add error message to require upgrading to 3.0 or 3.1 before any higher version

## 3.4.0

- mattermost 3.4.0
- add `MATTERMOST_SITE_URL` and `MATTERMOST_ENABLE_EMAIL_BATCHING` variables
- add `MATTERMOST_WEBSERVER_MODE` variable to control static file handling
- add `MATTERMOST_ENABLE_CUSTOM_EMOJI` variable
- add `MATTERMOST_RESTRICT_DIRECT_MESSAGE` variable

## 3.1.0

- mattermost 3.1.0
- add MATTERMOST_SERVER_LOCALE, MATTERMOST_CLIENT_LOCALE and MATTERMOST_LOCALES variables
- add MATTERMOST_MAX_FILE_SIZE

## 3.0.2

- mattermost 3.0.2
- add version 3.0 migration solution, see [README](README.md) for details
- remove `MATTERMOST_TEAM_DIRECTORY`
- add `MATTERMOST_OPEN_SERVER` to enable user creation without an invite
- add `MATTERMOST_PUSH_FULL_MESSAGE` to configure push notification content
- use `MATTERMOST_EMAIL_SIGNIN` and `MATTERMOST_USERNAME_SIGNIN` variables

## 2.2.0

- mattermost 2.2.0
- change base image to alpine:3.3
- remove embedded nginx server
- add support for postgresql
- add nginx container sample

## 2.1.0

- mattermost 2.1.0
- support gitlab integration
- support most mattermost configuration parameters


entrypoint.sh
#!/bin/bash
set -e

if [[ $DEBUG == true ]]; then
  set -x
  export MM_LOGSETTINGS_CONSOLE_LEVEL=DEBUG
  echo "======================"
  echo "Environment variables:"
  env
  echo "======================"
fi

#### Helper functions

vercmp() {
  # Compares two version strings `a` and `b`
  # Returns
  #   - negative integer, if `a` is less than `b`
  #   - 0, if `a` and `b` are equal
  #   - non-negative integer, if `a` is greater than `b`
  expr '(' "$1" : '\([^.]*\)' ')' '-' '(' "$2" : '\([^.]*\)' ')' '|' \
       '(' "$1.0" : '[^.]*[.]\([^.]*\)' ')' '-' '(' "$2.0" : '[^.]*[.]\([^.]*\)' ')' '|' \
       '(' "$1.0.0" : '[^.]*[.][^.]*[.]\([^.]*\)' ')' '-' '(' "$2.0.0" : '[^.]*[.][^.]*[.]\([^.]*\)' ')' '|' \
       '(' "$1.0.0.0" : '[^.]*[.][^.]*[.][^.]*[.]\([^.]*\)' ')' '-' '(' "$2.0.0.0" : '[^.]*[.][^.]*[.][^.]*[.]\([^.]*\)' ')'
}

configure_database() {
  echo "Configuring database..."
  finalize_database_parameters
  check_database_connection
  create_missing_database
}

finalize_database_parameters() {
  # is a mysql or postgresql named host available?
  echo -n "Trying to locate a database container."
  for ((i=0;i<10;i++)); do
    if ping -c1 -W1 mysql &> /dev/null; then
      MM_SQLSETTINGS_DRIVERNAME=${MM_SQLSETTINGS_DRIVERNAME:-mysql}
      DB_HOST=${DB_HOST:-mysql}
      echo -e "\u2714"
      break
    elif ping -c1 -W1 postgres &> /dev/null; then
      MM_SQLSETTINGS_DRIVERNAME=${MM_SQLSETTINGS_DRIVERNAME:-postgres}
      DB_HOST=${DB_HOST:-postgres}
      echo -e "\u2714"
      break
    fi
    echo -n "."
    sleep 1
  done

  # db agnostic defaults
  DB_NAME=${DB_NAME:-mattermost}

  # assemble datasource url
  case ${MM_SQLSETTINGS_DRIVERNAME} in
    mysql)
      DB_PORT=${DB_PORT:-3306}
      DB_USER=${DB_USER:-root}
      DB_ENCODING=${DB_ENCODING:-utf8mb4,utf8}
      DB_PARAMS="charset=${DB_ENCODING}"
      export MM_SQLSETTINGS_DATASOURCE="${DB_USER}:${DB_PASS}@tcp(${DB_HOST}:${DB_PORT})/${DB_NAME}?${DB_PARAMS}"
      ;;
    postgres)
      DB_PORT=${DB_PORT:-5432}
      DB_USER=${DB_USER:-postgres}
      DB_PARAMS="sslmode=disable&connect_timeout=10"
      export MM_SQLSETTINGS_DATASOURCE="postgres://${DB_USER}:${DB_PASS}@${DB_HOST}:${DB_PORT}/${DB_NAME}?${DB_PARAMS}"
      ;;
    *)
      echo
      echo "ERROR: Supported values for MM_SQLSETTINGS_DRIVERNAME are 'postgres' and 'mysql'."
      echo
      exit 1
      ;;
  esac

}

check_database_connection() {
  local check

  echo -n "Checking database connection"
  case ${MM_SQLSETTINGS_DRIVERNAME} in
    mysql)
      check=(mysqladmin -h ${DB_HOST} -P ${DB_PORT} -u ${DB_USER} -p${DB_PASS} status)
      ;;
    postgres)
      export PGPASSWORD=${DB_PASS}
      check=(psql -h ${DB_HOST} -p ${DB_PORT} -U ${DB_USER})
      ;;
  esac

  connectable=1
  for ((i=0;i<60;i++)); do
    if "${check[@]}" &> /dev/null; then
      connectable=0
      break
    fi
    echo -n "."
    sleep 1
  done

  if [[ $connectable -eq 1 ]]; then
    echo
    echo "Could not connect to database server. Aborting..."
    exit 1
  else
    echo -e "\u2714"
  fi

}

create_missing_database() {
  echo -n "Testing whether database ${DB_NAME} exists. "

  case ${MM_SQLSETTINGS_DRIVERNAME} in
    mysql)
      echo -n "Creating database ${DB_NAME} if necessary. "
      mysql -h ${DB_HOST} -P ${DB_PORT} -u ${DB_USER} -p${DB_PASS} mysql << EOF
        CREATE DATABASE IF NOT EXISTS ${DB_NAME};
        GRANT ALL PRIVILEGES ON ${DB_NAME} TO ${DB_USER};
EOF
      echo -e "\u2714"
      ;;
    postgres)
      if psql ${MM_SQLSETTINGS_DATASOURCE}; then
        echo -e "\u2714"
      else
        echo -e "\u2718"
        echo -n "Creating database ${DB_NAME}. "
        psql -h ${DB_HOST} -p ${DB_PORT} -U ${DB_USER} << EOF
          CREATE DATABASE ${DB_NAME} WITH OWNER ${DB_USER};
          GRANT ALL PRIVILEGES ON ${DB_NAME} TO ${DB_USER};
EOF
        echo -e "\u2714"
      fi
      ;;

  esac
}

check_version() {
  if [ -f ${MATTERMOST_DATA_DIR}/VERSION ]; then
    CACHE_VERSION=$(cat ${MATTERMOST_DATA_DIR}/VERSION)
  else
    CACHE_VERSION=${MATTERMOST_VERSION}
  fi

  if [[ $(vercmp ${MATTERMOST_VERSION} ${CACHE_VERSION}) -lt 0 ]]; then
    echo
    echo "ERROR: "
    echo "  Cannot downgrade from Mattermost version ${CACHE_VERSION} to ${MATTERMOST_VERSION}."
    echo "  Only upgrades are allowed. Please use jasl8r/mattermost:${CACHE_VERSION} or higher."
    echo "  Cannot continue. Aborting!"
    echo
    exit 1
  fi

  if [[ $(vercmp ${CACHE_VERSION} 3.5.1) -lt 0 ]]; then
    echo
    echo "ERROR: "
    echo "  Cannot upgrade from Mattermost version ${CACHE_VERSION} to ${MATTERMOST_VERSION}."
    echo "  Mattermost version ${CACHE_VERSION} must be upgraded to version 3.5.1 first."
    echo "  Please run jasl8r/mattermost:3.5.1 followed by funkyfuture/mattermost:${MATTERMOST_VERSION}"
    echo "  Cannot continue. Aborting!"
    echo
    exit 1
  fi

  echo "${MATTERMOST_VERSION}" > ${MATTERMOST_DATA_DIR}/VERSION
}

#### Main procedure

# ensure file access
chmod 750 ${MATTERMOST_DATA_DIR}
chown -R "$(id -u):$(id -g)" ${MATTERMOST_DATA_DIR}

# database configuration
check_version
configure_database

for setting_variable in ${!MM_*}; do
  eval "export ${setting_variable}"
done

exec ./bin/platform --config ${MATTERMOST_CONF_DIR}/config.json "$@"

exit 1
